### PR TITLE
Go Structs for GEP-91

### PR DESCRIPTION
**Pull request by arkodg**
/kind documentation
/kind feature
/kind gep

**What this PR does / why we need it**:

Takes https://github.com/kubernetes-sigs/gateway-api/pull/2759 forward and implements the go structs described in GEP-91 https://gateway-api.sigs.k8s.io/geps/gep-91/

**Which issue(s) this PR fixes**:

Fixes #

**Does this PR introduce a user-facing change?**:

```release-note
Adds support for a new FrontendValidation field with the Gateway Listener TLS field to support validation of client certificates using a CA Certificate as a trust anchor
```

### CODE REVIEW

**Comment by arkodg** [comment]
cc @robscott

File: config/crd/experimental/gateway.networking.k8s.io_gateways.yaml [hunk 1]
```diff
@@ -720,6 +720,104 @@ spec:
                             type: object
                           maxItems: 64
                           type: array
+                        frontendValidation:
```
Comment (robscott) [inline config/crd/experimental/gateway.networking.k8s.io_gateways.yaml:723]: Can you add an example to `/examples/experimental` that uses this field to ensure that it's all working as intended?

**Review by robscott — COMMENTED** [review]
Thanks @arkodg! This mostly LGTM, just looks like some codegen issues in presubmit.

**Comment by arkodg** [comment]
not seeing any failures when running `./hack/verify-all.sh -v` locally
except for this one, unsure how its related to this PR
```
Verifying docker images
./hack/../hack/verify-docker-build.sh: line 25: BUILDX_CONTEXT: unbound variable
Test FAILED: ./hack/../hack/verify-docker-build.sh 
```

**Comment by robscott** [comment]
@arkodg I tried running hack/update-codegen.sh from your PR and ended up with changes to a couple generated GRPC pb files:

```
diff --git a/conformance/echo-basic/grpcechoserver/grpcecho.pb.go b/conformance/echo-basic/grpcechoserver/grpcecho.pb.go
index 81135b30..9585b776 100644
--- a/conformance/echo-basic/grpcechoserver/grpcecho.pb.go
+++ b/conformance/echo-basic/grpcechoserver/grpcecho.pb.go
@@ -14,7 +14,7 @@

 // Code generated by protoc-gen-go. DO NOT EDIT.
 // versions:
-//     protoc-gen-go v1.28.0
+//     protoc-gen-go v1.28.1
 //     protoc        v4.22.2
 // source: grpcecho.proto

diff --git a/conformance/echo-basic/grpcechoserver/grpcecho_grpc.pb.go b/conformance/echo-basic/grpcechoserver/grpcecho_grpc.pb.go
index 67d46680..f367cb66 100644
--- a/conformance/echo-basic/grpcechoserver/grpcecho_grpc.pb.go
+++ b/conformance/echo-basic/grpcechoserver/grpcecho_grpc.pb.go
@@ -1,4 +1,8 @@
 // Code generated by protoc-gen-go-grpc. DO NOT EDIT.
+// versions:
+// - protoc-gen-go-grpc v1.2.0
+// - protoc             v4.22.2
+// source: grpcecho.proto

 package grpcechoserver

@@ -11,6 +15,7 @@ import (

 // This is a compile-time assertion to ensure that this generated file
 // is compatible with the grpc package it is being compiled against.
+// Requires gRPC-Go v1.32.0 or later.
 const _ = grpc.SupportPackageIsVersion7

 // GrpcEchoClient is the client API for GrpcEcho service.
@@ -95,8 +100,8 @@ type UnsafeGrpcEchoServer interface {
        mustEmbedUnimplementedGrpcEchoServer()
 }

-func RegisterGrpcEchoServer(s *grpc.Server, srv GrpcEchoServer) {
-       s.RegisterService(&_GrpcEcho_serviceDesc, srv)
+func RegisterGrpcEchoServer(s grpc.ServiceRegistrar, srv GrpcEchoServer) {
+       s.RegisterService(&GrpcEcho_ServiceDesc, srv)
 }

 func _GrpcEcho_Echo_Handler(srv interface{}, ctx context.Context, dec func(interface{}) error, interceptor grpc.UnaryServerInterceptor) (interface{}, error) {
@@ -153,7 +158,10 @@ func _GrpcEcho_EchoThree_Handler(srv interface{}, ctx context.Context, dec func(
        return interceptor(ctx, in, info, handler)
 }

-var _GrpcEcho_serviceDesc = grpc.ServiceDesc{
+// GrpcEcho_ServiceDesc is the grpc.ServiceDesc for GrpcEcho service.
+// It's only intended for direct use with grpc.RegisterService,
+// and not to be introspected or modified (even as a copy)
+var GrpcEcho_ServiceDesc = grpc.ServiceDesc{
        ServiceName: "gateway_api_conformance.echo_basic.grpcecho.GrpcEcho",
        HandlerType: (*GrpcEchoServer)(nil),
        Methods: []grpc.MethodDesc{
~
~
~
(END)...skipping...
diff --git a/conformance/echo-basic/grpcechoserver/grpcecho.pb.go b/conformance/echo-basic/grpcechoserver/grpcecho.pb.go
index 81135b30..9585b776 100644
--- a/conformance/echo-basic/grpcechoserver/grpcecho.pb.go
+++ b/conformance/echo-basic/grpcechoserver/grpcecho.pb.go
@@ -14,7 +14,7 @@

 // Code generated by protoc-gen-go. DO NOT EDIT.
 // versions:
-//     protoc-gen-go v1.28.0
+//     protoc-gen-go v1.28.1
 //     protoc        v4.22.2
 // source: grpcecho.proto

diff --git a/conformance/echo-basic/grpcechoserver/grpcecho_grpc.pb.go b/conformance/echo-basic/grpcechoserver/grpcecho_grpc.pb.go
index 67d46680..f367cb66 100644
--- a/conformance/echo-basic/grpcechoserver/grpcecho_grpc.pb.go
+++ b/conformance/echo-basic/grpcechoserver/grpcecho_grpc.pb.go
@@ -1,4 +1,8 @@
 // Code generated by protoc-gen-go-grpc. DO NOT EDIT.
+// versions:
+// - protoc-gen-go-grpc v1.2.0
+// - protoc             v4.22.2
+// source: grpcecho.proto

 package grpcechoserver

@@ -11,6 +15,7 @@ import (

 // This is a compile-time assertion to ensure that this generated file
 // is compatible with the grpc package it is being compiled against.
+// Requires gRPC-Go v1.32.0 or later.
 const _ = grpc.SupportPackageIsVersion7

 // GrpcEchoClient is the client API for GrpcEcho service.
@@ -95,8 +100,8 @@ type UnsafeGrpcEchoServer interface {
        mustEmbedUnimplementedGrpcEchoServer()
 }

-func RegisterGrpcEchoServer(s *grpc.Server, srv GrpcEchoServer) {
-       s.RegisterService(&_GrpcEcho_serviceDesc, srv)
+func RegisterGrpcEchoServer(s grpc.ServiceRegistrar, srv GrpcEchoServer) {
+       s.RegisterService(&GrpcEcho_ServiceDesc, srv)
 }

 func _GrpcEcho_Echo_Handler(srv interface{}, ctx context.Context, dec func(interface{}) error, interceptor grpc.UnaryServerInterceptor) (interface{}, error) {
@@ -153,7 +158,10 @@ func _GrpcEcho_EchoThree_Handler(srv interface{}, ctx context.Context, dec func(
        return interceptor(ctx, in, info, handler)
 }

-var _GrpcEcho_serviceDesc = grpc.ServiceDesc{
+// GrpcEcho_ServiceDesc is the grpc.ServiceDesc for GrpcEcho service.
+// It's only intended for direct use with grpc.RegisterService,
+// and not to be introspected or modified (even as a copy)
+var GrpcEcho_ServiceDesc = grpc.ServiceDesc{
        ServiceName: "gateway_api_conformance.echo_basic.grpcecho.GrpcEcho",
        HandlerType: (*GrpcEchoServer)(nil),
        Methods: []grpc.MethodDesc{
```

I think that would resolve the failing presubmit, I'm guessing you're not seeing the same diff when you run codegen?

**Comment by arkodg** [comment]
ha thanks @robscott !

bumping `protoc-gen-go` & `protoc-gen-go-grpc` did the trick 

```
go install google.golang.org/protobuf/cmd/protoc-gen-go@v1.28
go install google.golang.org/grpc/cmd/protoc-gen-go-grpc@v1.2
  ```
  
do have a philosophy on how we pin dev tools in this project ? for e.g. in some projects these tools are prefetched and saved  in a local `tools/bin` dir to ensure tooling consistency (and use dependabot to keep bumping them ) for all,

**Review by shaneutt — COMMENTED** [review]
After existing comments are resolved, this generally LGTM to me.

/lgtm

But hold just so it doesn't immediately merge.

/hold

**Comment by robscott** [comment]
@arkodg looks like another PR beat this one in, will LGTM once you rebase.

/hold cancel

**Pull request merged**